README.md

```diff
@@ -1,251 +1,395 @@
+## Enumerating Dictionaries
 
-Again, because this syntax is more verbose, we don't suggest you default to it, but you may come across it, and therefore you must be familiar with it.
+We can access the entire contents of a dictionary sequentially by iterating over it with a `for-in` loop. The example syntax tells the compiler to treat each key as a string:
 
-## `alloc` and `init`
+```objc
+for (NSString *key in dictionary) {
+    Class *element = dictionary[key];
+}
+```
 
-As with all objects in Objective-C, you may also use the more traditional `alloc` and `init` to initialize a dictionary, like so:
+So, we can use a `for-in` loop to print out each of the details we have for Jenny:
 
 ```objc
-NSDictionary *marcDictionary = [[NSDictionary alloc] initWithObjectsAndKeys:@2125551212, @"Phone Number", nil]`
+NSDictionary *jenny = @{ @"first_name"       : @"Jenny",
+                         @"relationship"     : @"Friend",
+                         @"phone_number"     : @"(555) 867-5309",
+                         @"email_address"    : @"jenny@email.com",
+                         @"physical_address" : @"123 Street Name",
+                         @"city_state"       : @"Anywhere, USA",
+                         @"zip_code"         : @"00409"        };
+    
+for (NSString *key in jenny) {
+    NSLog(@"Jenny's %@ is %@.", key, jenny[key]);
+}
 ```
+This will print: 
 
-Like `-objectForKey`, this syntax is old, verbose, error-prone (you *must* add `nil` at the end), and somewhat confusing (values come before their keys?!). There is also a method `-initWithObjects:forKeys:` that matches up corresponding elements of two arrays. It is equally verbose but arguably less error-prone.
-
-Use the literal syntax, but be aware that you may see these methods in older code.
+```
+Jenny's email_address is jenny@email.com.
+Jenny's zip_code is 00409.
+Jenny's physical_address is 123 Street Name.
+Jenny's relationship is Friend.
+Jenny's phone_number is (555) 867-5309.
+Jenny's city_state is Anywhere, USA.
+Jenny's first_name is Jenny.
+```
+Notice again that the order was not preserved.
 
+#### Grabbing `allKeys`
 
-## Enumerating Dictionaries
+![](https://curriculum-content.s3.amazonaws.com/ios/reading-ios-nsdictionary/grabAllKeys_meme.jpg)
 
-You have already learned how to use a `for..in` loop. `for..in` loops are defined to loop over the keys of `NSDictionary` objects, so we can run through all of the phone numbers in our contact list like so:
+There's a handy method on `NSDictionary` called `allKeys` that returns an array of all of the keys (and only the keys). This is helpful in sorting and searching (which we'll discuss in other readings), but also in defensive programming to make sure that a dictionary will respond to a key before attempting to access it. Implementing a "key check" can avoid receiving `nil` from attempts to access a dictionary key that doesn't exist:
```

Comment:
I would specifically point out that requesting a key that doesn't exist returns a `nil`